ext/js/language/ja/text-preprocessors.js

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2024  Yomitan Authors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import {basicTextPreprocessorOptions} from '../text-preprocessors.js';
+import {convertAlphabeticToKana} from './japanese-wanakana.js';
+import {collapseEmphaticSequences, convertHalfWidthKanaToFullWidth, convertHiraganaToKatakana, convertKatakanaToHiragana, convertNumericToFullWidth} from './japanese.js';
+
+/** @type {import('language').TextPreprocessor<[collapseEmphatic: boolean, collapseEmphaticFull: boolean]>}*/
+const collapseEmphaticSequencesPrerocessor = {
+    id: 'collapseEmphaticSequences',
+    name: 'Collapse emphatic character sequences',
+    description: 'すっっごーーい → すっごーい / すごい',
+    options: [[false, false], [true, false], [true, true]],
+    process: (str, setting, sourceMap) => {
+        const [collapseEmphatic, collapseEmphaticFull] = setting;
+        if (collapseEmphatic) {
+            str = collapseEmphaticSequences(str, collapseEmphaticFull, sourceMap);
+        }
+        return str;
+    }
+};
+
+/** @type {import('language').TextPreprocessor<any>[]} */
```

Comment:
This is the main thing that I think could use better typing due to the use of `any`. At a minimum, it needs to use `SafeAny` to show that we explicitly evaluated this, but I also still wonder if we can do something here like:

`import('language').TextPreprocessorAny`

where

`type TextPreprocessorAny = TextPreprocessor<A> | TextPreprocessor<B> | ...`

something like that.